src/common/ThreadLocalPrebufferedStreambuf.h

```diff
@@ -0,0 +1,83 @@
+#ifndef CEPH_COMMON_THREAD_LOCAL_PREBUFFEREDSTREAMBUF_H
+#define CEPH_COMMON_THREAD_LOCAL_PREBUFFEREDSTREAMBUF_H
+
+#include <streambuf>
+#include <atomic>
+#include <ostream>
+/**
+ * streambuf using existing buffer, overflowing into a std::string
+ *
+ * A simple streambuf that uses a preallocated buffer for small
+ * strings, and overflows into a std::string when necessary.  If the
+ * preallocated buffer size is chosen well, we can optimize for the
+ * common case and overflow to a slower heap-allocated buffer when
+ * necessary.
+ */
+
+struct prebuffered_data
+{
+private:
+  char *m_buf;
+  size_t m_buf_len;
+  char *m_pptr;
+  std::string m_overflow;
+public:
+  prebuffered_data(char* buf, size_t buf_len)
+  : m_buf(buf), m_buf_len(buf_len), m_pptr(nullptr) {}
+
+  /// return a string copy (inefficiently)
+  std::string get_str() const;
```

Comment:
please see https://github.com/ceph/ceph/pull/19100/files#diff-8dcb74089b1227130d768a47852508bdR48, `get_str()`, `size()` and `snprintf()` methods are not needed here.